File: switsol_ag/custom_scripts/address.js
// Copyright (c) 2017, Switsol AG and contributors
// For license information, please see license.txt
// cur_frm.add_fetch("postal_code", "canton", "canton");
cur_frm.add_fetch("town", "canton", "canton")

frappe.ui.form.on("Address", {
	onload: function(frm) {
		frm.set_query("country", function() {
			return {
				filters: {
					"flag": 1
				}
			};
		});
		frm.set_query("postal_code", function(doc, cdt, cdn) {
			if(frm.doc.town) {
				return {
					query: "switsol_ag.address.postal_code_query",
					filters: {
						"town": frm.doc.town
					}
				}
			}
		});
		frm.set_query("town", function(doc, cdt, cdn) {
			if(frm.doc.postal_code) {
				return {
					query: "switsol_ag.address.city_query",
					filters: {
						"postal_code": frm.doc.postal_code
					}
				}
			}
		});
	},
	refresh: function(frm) {
		if(frappe.route_options) {
			cur_frm.set_value("address_title", frappe.route_options.address_title);
			cur_frm.set_value("address_type", "Billing");
			cur_frm.set_value("country", frappe.sys_defaults.country);
		}
	},
	town: function(frm) {
		frappe.call({
			method: "switsol_ag.address.get_list_switsol",
			args: {
				doctype: "City Postal Code",
				filters: [
					["parent", "=", frm.doc.town]
				],
				fields: ["postal_code"]
			},
			callback: function(r) {
				if(r.message) {
					let postal_codes = [];
					$.each(r.message, function(j, data) {
						postal_codes.push(data.postal_code);
					})
					if(postal_codes.length == 1) {
						frm.set_value("postal_code", postal_codes[0]);
					}
				}
			}
		})
	}
});

File: switsol_ag/address.py
# -*- coding: utf-8 -*-
# Copyright (c) 2017, Switsol AG and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe


def validate(self, method=None):
    self.pincode = self.postal_code.strip() if self.postal_code else ""
    # if self.customer and self.is_primary_address:
    #     frappe.db.set_value("Customer", self.customer, "pincode", self.postal_code)


@frappe.whitelist()
def postal_code_query(doctype, txt, searchfield, start, page_len, filters):
    return frappe.db.sql("""
        SELECT
            postal_code
        FROM
            `tabCity Postal Code`
        WHERE
            parent=%s
        AND 
            postal_code LIKE %s
        ORDER BY postal_code DESC
        LIMIT %s, %s""".format(frappe.db.escape(searchfield)), (filters.get("town"), "%{0}%".format(txt), start, page_len))


@frappe.whitelist()
def city_query(doctype, txt, searchfield, start, page_len, filters):
    return frappe.db.sql("""
        SELECT
            cp.parent
        FROM
            `tabCity Postal Code` cp
        WHERE
            cp.postal_code=%s
        ORDER BY parent DESC
        LIMIT %s, %s""".format(frappe.db.escape(searchfield)), (filters.get("postal_code"), start, page_len))


@frappe.whitelist()
def make_address(company_name):
    ad = frappe.new_doc("Address")
    ad.customer = company_name
    ad.address_title = company_name
    return ad


@frappe.whitelist()
def get_list_switsol(doctype, fields=None, filters=None, order_by=None,
    limit_start=None, limit_page_length=20, parent=None):
    '''Returns a list of records by filters, fields, ordering and limit

    :param doctype: DocType of the data to be queried
    :param fields: fields to be returned. Default is `name`
    :param filters: filter list by this dict
    :param order_by: Order by this fieldname
    :param limit_start: Start at this index
    :param limit_page_length: Number of records to be returned (default 20)'''
    return frappe.get_list(doctype, fields=fields, filters=filters, order_by=order_by,
        limit_start=limit_start, limit_page_length=limit_page_length, ignore_permissions=False)
